chainntnfs/height_hint_cache.go

```diff
@@ -0,0 +1,303 @@
+package chainntnfs
+
+import (
+	"bytes"
+	"errors"
+	"os"
+	"path/filepath"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	bolt "github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/channeldb"
+)
+
+const (
+	// dbName is the default name of the database storing the height hints.
+	dbName = "heighthint.db"
+
+	// dbFilePermission is the default permission of the database file
+	// storing the height hints.
+	dbFilePermission = 0600
+)
+
+var (
+	// spendHintBucket is the name of the bucket which houses the height
+	// hint for outpoints. Each height hint represents the earliest height
+	// at which its corresponding outpoint could have been spent within.
+	spendHintBucket = []byte("spend-hints")
+
+	// confirmHintBucket is the name of the bucket which houses the height
+	// hints for transactions. Each height hint represents the earliest
+	// height at which its corresponding transaction could have been
+	// confirmed within.
+	confirmHintBucket = []byte("confirm-hints")
+
+	// ErrCorruptedHeightHintCache indicates that the on-disk bucketing
+	// structure has altered since the height hint cache instance was
+	// initialized.
+	ErrCorruptedHeightHintCache = errors.New("height hint cache has been " +
+		"corrupted")
+
+	// ErrSpendHintNotFound is an error returned when a spend hint for an
+	// outpoint was not found.
+	ErrSpendHintNotFound = errors.New("spend hint not found")
+
+	// ErrConfirmHintNotFound is an error returned when a confirm hint for a
+	// transaction was not found.
+	ErrConfirmHintNotFound = errors.New("confirm hint not found")
+)
+
+// HeightHintCache is responsible for caching the height hints for outpoints and
+// transactions. A height hint for an outpoint signifies the earliest height at
+// which this outpoint could have been spent within. A height hint for a
+// transaction signifies the earliest height at which this transaction could
+// have been included within a block.
+type HeightHintCache interface {
+	// CommitSpendHint commits a spend hint for the outpoints to the cache.
+	CommitSpendHint(heightHint uint32, ops ...wire.OutPoint) error
+
+	// QuerySpendHint returns the latest spend hint for an outpoint.
+	// ErrSpendHintNotFound is returned if a spend hint does not exist
+	// within the cache for the outpoint.
+	QuerySpendHint(op wire.OutPoint) (uint32, error)
+
+	// PurgeSpendHint removes the spend hint for the outpoints from the
+	// cache.
+	PurgeSpendHint(ops ...wire.OutPoint) error
```

Comment:
I don't see hints purged _anywhere_ in the PR as is. It seems safe to just use our regular re-org horizon. Since we now have _persistent_ spend notifications for each channel via the contract court, we no longer run into an edge case of a output being spent, but peer not coming online for hundreds of blocks leading us to remove a stale height hint entry.

Reply:
> It seems safe to just use our regular re-org horizon.

Due to the way the `TxConfNotifier` works, it's possible that we end up not removing some hints from the cache if we go down before some transactions reach the reorg horizon.

> Since we now have persistent spend notifications for each channel via the contract court, we no longer run into an edge case of a output being spent, but peer not coming online for hundreds of blocks leading us to remove a stale height hint entry.

Not sure what you mean by "persistent spend notifications". Isn't this what the cache aims to solve?